File: examples/python/leetcode75/isomorphicString.py
# 205. Isomorphic Strings
# https://leetcode.com/problems/isomorphic-strings


def isIsomorphic(s: str, t: str) -> bool:
    if len(s) != len(t):
        return False

    d1, d2 = {}, {}

    for i in range(len(s)):
        if s[i] not in d1.keys():
            d1[s[i]] = t[i]
        if s[i] in d1.keys():
            if t[i] != d1[s[i]]:
                return False

    for i in range(len(t)):
        if t[i] not in d2.keys():
            d2[t[i]] = s[i]
        if t[i] in d2.keys():
            if s[i] != d2[t[i]]:
                return False

    if d1 == dict((v, k) for k, v in d2.items()):
        return True

    return False


if __name__ == "__main__":
    # s = "egg"
    # t = "add"

    s = "abbaaab"
    t = "baabbab"

    print(isIsomorphic(s, t))

File: LC_problems/205/isomorphicString.cpp
// 205. Isomorphic Strings
// https://leetcode.com/problems/isomorphic-strings

#include <iostream>
#include <string>
#include <unordered_map>

bool isIsomorphic(std::string s, std::string t)
{
    std::unordered_map<char, char> m;
    std::unordered_map<char, char> m2;

    for (int i = 0; i < s.size(); i++)
    {
        if (m.find(s[i]) == m.end())
            m[s[i]] = t[i];
        else if (m[s[i]] != t[i])
            return false;
    }

    for (int i = 0; i < t.size(); i++)
    {
        if (m2.find(t[i]) == m2.end())
            m2[t[i]] = s[i];
        else if (m2[t[i]] != s[i])
            return false;
    }

    return true;
}

int main()
{
    std::string s = "badc";
    std::string t = "baba";

    bool isIso;

    isIso = isIsomorphic(s, t);

    if (isIso)
        std::cout << "These strings are Isomorphic";
    else
        std::cout << "These strings are not Isomorphic";
}